src/DFAMin.hs

```diff
@@ -112,57 +131,48 @@ groupEquivStates DFA { dfa_states = statemap }
     accept_groups :: [EquivalenceClass]
     accept_groups = map IS.fromList (Map.elems accept_map)
 
-    init_p, init_q :: [EquivalenceClass]
-    init_p  -- Issue #71: each EquivalenceClass needs to be a non-empty set
-      | IS.null nonaccepting_states = accept_groups
-      | otherwise                   = nonaccepting_states : accept_groups
+    init_r, init_q :: [EquivalenceClass]
+    init_r  -- Issue #71: each EquivalenceClass needs to be a non-empty set
+      | IS.null nonaccepting_states = []
+      | otherwise                   = [nonaccepting_states]
     init_q = accept_groups
 
-    -- map token T to
-    --   a map from state S to the list of states that transition to
-    --   S on token T
-    -- This is a cache of the information needed to compute x below
-    bigmap :: IntMap (IntMap [SNum])
-    bigmap = IM.fromListWith (IM.unionWith (++))
-                [ (i, IM.singleton to [from])
+    -- elements of
+    --   a map from token T to
+    --     a map from state S to the set of states that transition to
+    --     S on token T
+    -- This is a cache of the information needed to compute xs below
+    bigmap :: [IntMap EquivalenceClass]
+    bigmap = IM.elems $ IM.fromListWith (IM.unionWith IS.union)
```

Comment:
Can you take out the `IM.elems`? Yes, I know you want to deduplicate that, but that change alone is a very basic common sub-exrepssion elimination that GHC has no trouble with. Conversely, I find the *meaning* of a single big map vs list of sets easier to understand, even if we don't need the map structure

Reply:
I decided to trust GHC XD
https://github.com/simonmar/alex/pull/176/commits/c9bc39ab9f19c2e1b045f898fb21828e4781831f